[PluggableDevice] Add pluggable profiler C Interface support

jzhoulon opened the pull request:
Add TensorFlow profiler support for PluggableDevice.

jzhoulon commented:
@penpornk @yisitu @jbaiocchi This PR is the implementation of the RFC: [Add pluggable device profiler support ](https://github.com/tensorflow/community/pull/389), please help to review. Thanks very much

jzhoulon committed:
[PluggableDevice] Add pluggable profiler support

jbaiocchi commented on tensorflow/c/experimental/profiler/profiler.cc:
```diff
@@ -0,0 +1,225 @@
+    return absl::WrapUnique(new PluggableProfiler(profiler_fns, profiler));
+  }
+
+  Status Start() {
```
Add 'override' to functions that implement ProfilerInterface.
https://google.github.io/styleguide/cppguide.html#Inheritance

jzhoulon replied:
done

jbaiocchi commented on tensorflow/c/experimental/profiler/profiler.cc:
```diff
@@ -0,0 +1,225 @@
+    return tensorflow::StatusFromTF_Status(c_status.get());
+  }
+
+  Status CollectData(RunMetadata* run_metadata) {
```
This function can be removed, RunMetadata overload of CollectData has been removed from ProfilerInterface.

jzhoulon replied:
Done

jbaiocchi commented on tensorflow/c/experimental/profiler/profiler.cc:
```diff
@@ -0,0 +1,225 @@
+
+  Status CollectData(XSpace* space) {
+    std::unique_ptr<TF_Status, TFStatusDeleter> c_status(TF_NewStatus());
+    // Get size of buffer required for Plugin to serialize XSpace ito.
```
into it ?

jzhoulon replied:
done

jbaiocchi commented on tensorflow/c/experimental/profiler/profiler.cc:
```diff
@@ -0,0 +1,225 @@
+        return factory.CreatePluggableProfiler(options);
+      };
+
+  tensorflow::profiler::RegisterProfilerFactory(create_func);
```
std::move(create_func)

jzhoulon replied:
done

jbaiocchi commented on tensorflow/c/experimental/profiler/BUILD:
```diff
@@ -0,0 +1,64 @@
+)
+
+cc_library(
+    name = "profiler_hdrs",
```
Can you rename this directory and all the files to be "pluggable_profiler" instead of just "profiler"?

jzhoulon replied:
done

jbaiocchi commented on tensorflow/c/experimental/profiler/profiler.cc:
```diff
@@ -0,0 +1,225 @@
+namespace tensorflow {
+namespace profiler {
+
+using OwnedTFStatus = std::unique_ptr<TF_Status, TFStatusDeleter>;
```
Is this used? If not remove.

jzhoulon replied:
done

jbaiocchi commented on tensorflow/c/experimental/profiler/profiler_internal.h:
```diff
@@ -0,0 +1,35 @@
+#include "tensorflow/core/profiler/protobuf/xplane.pb.h"
+namespace tensorflow {
+namespace profiler {
+struct TFStatusDeleter {
```
This can be removed.

jzhoulon replied:
done

jbaiocchi commented on tensorflow/c/experimental/profiler/profiler.cc:
```diff
@@ -0,0 +1,225 @@
+  }
+
+  Status Start() {
+    std::unique_ptr<TF_Status, TFStatusDeleter> c_status(TF_NewStatus());
```
The pattern I see elsewhere in the code base is:

std::unique_ptr<TF_Status, decltype(&TF_DeleteStatus)> status(
      TF_NewStatus(), TF_DeleteStatus);

I would also name this variable status instead of c_status, the prefix does not add much value in this case.

jzhoulon replied:
done

jbaiocchi commented on tensorflow/c/experimental/profiler/profiler_internal.h:
```diff
@@ -0,0 +1,35 @@
+  void operator()(TF_Status* s) const { TF_DeleteStatus(s); }
+};
+
+typedef void (*TFInitProfilerFn)(TF_ProfilerRegistrationParams* const,
```
Rewrite as "using TFInitProfilerFn = ".
https://google.github.io/styleguide/cppguide.html#Aliases

jzhoulon replied:
done

jbaiocchi commented on tensorflow/c/experimental/profiler/profiler_internal.h:
```diff
@@ -0,0 +1,35 @@
+typedef void (*TFInitProfilerFn)(TF_ProfilerRegistrationParams* const,
+                                 TF_Status* const);
+
+Status InitPluginProfiler(TFInitProfilerFn init_fn);
```
Add a black-box documentation comment to explain what this function does.

jzhoulon replied:
- done

jbaiocchi commented on tensorflow/c/experimental/profiler/profiler.h:
```diff
@@ -0,0 +1,92 @@
+extern "C" {
+#endif
+
+typedef struct TP_Profiler {
```
Add documentation comments for each of the structs.

jzhoulon replied:
done

jbaiocchi commented on tensorflow/core/common_runtime/pluggable_device/pluggable_device_plugin_init.cc:
```diff
@@ -81,13 +82,26 @@ static Status InitKernelModule(void* dso_handle) {
   TF_RETURN_IF_ERROR(InitKernelModule(dso_handle));
 
+  // Step 3 Init Profiler Module(optional module)
+  InitProfilerModule(dso_handle);
```
A returned Status must not be ignored. If not using TF_RETURN_IF_ERROR, explicity call IgnoreError() on the result of the call and explain in a comment why it is OK to ignore. Even better, log the resulting status so for ease of debugging.

jzhoulon replied:
Done

jzhoulon commented:
@jbaiocchi Thanks for the comments, I have addressed all the comments, please help review it. thanks very much

jbaiocchi commented on tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.h:
```diff
@@ -0,0 +1,169 @@
+//   /* Sample TensorFlow code below, exact implementation might differ. */
+//   // Version checking uses `struct_size`. It is exempt from the `TF/TP` rule
+//   // above and should be set both by core and the plugin."
+
```
nit: add // even to these blank lines as for a single comment block

jbaiocchi commented on tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.cc:
```diff
@@ -0,0 +1,223 @@
+#define VALIDATE_STRUCT_SIZE(STRUCT_NAME, STRUCT_OBJ, SIZE_VALUE_NAME) \
+  do {                                                                 \
+    if (STRUCT_OBJ.struct_size == 0) {                                 \
+      return Status(error::FAILED_PRECONDITION,                        \
+                    "struct_size field in " #STRUCT_NAME               \
+                    " must be set to " #SIZE_VALUE_NAME ".");          \
+    }                                                                  \
+  } while (0)
+
+#define VALIDATE_MEMBER(STRUCT_NAME, STRUCT_OBJ, NAME)                     \
+  do {                                                                     \
+    if (STRUCT_OBJ.NAME == 0) {                                            \
+      return Status(error::FAILED_PRECONDITION,                            \
+                    "'" #NAME "' field in " #STRUCT_NAME " must be set."); \
+    }                                                                      \
+  } while (0)
+
+Status ValidateDeviceType(StringPiece type) {
+  // Validate device type. Device type must start with a capital letter and
+  // consist of capital letters and underscores. Reasoning behind this decision:
+  // * At the minimum we want to disallow '/' and ':' since
+  //   these characters are used in device spec, for e.g.
+  //   /job:foo/replica:12/device:GPU:1.
+  // * Underscores seem useful, for e.g. XLA_GPU uses underscores.
+  // * Allowing lowercase might get confusing. For example, say someone
+  //   registers a new type called "Gpu". It might be confusing for users that
+  //   "Gpu" is not the same device type as "GPU".
+  //   Note that lowercase "cpu" and "gpu" are currently supported only for
+  //   legacy reasons:
+  //   https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/python/framework/device_spec.py;l=46;drc=d3a378f9665d8eee827c74cb9ecbee81e4c288dd
+  static const LazyRE2 kTfDeviceTypeRegEx = {"[A-Z][A-Z_]*"};
+  bool matches = RE2::FullMatch(type, *kTfDeviceTypeRegEx);
+  if (!matches) {
+    return Status(
+        error::FAILED_PRECONDITION,
+        tensorflow::strings::StrCat("Device name/type '", type, "' must match ",
+                                    kTfDeviceTypeRegEx->pattern(), "."));
+  }
+  return Status::OK();
+}
```
This is duplicated code from tensorflow/c/experimental/stream_executor/stream_executor.cc
Can we prevent duplication by first extracting a common library, e.g., pluggable_device_utils ?

jbaiocchi commented on tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.cc:
```diff
@@ -0,0 +1,223 @@
+
+struct TFStatusDeleter {
+  void operator()(TF_Status* s) const { TF_DeleteStatus(s); }
+};
```
This struct and the OwnedTFStatus alias are also used in third_party/tensorflow/c/experimental/stream_executor/stream_executor.cc

Move them to common utilities as well.

jbaiocchi reviewed (changes requested):
Please address the code duplication issue.

jzhoulon committed:
address comments

jzhoulon committed:
address comments

penpornk commented on tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.h:
```diff
@@ -0,0 +1,44 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
```
Let's wrap functions here in `namespace pluggable_device` to avoid symbol clashes.

jzhoulon replied:
done

penpornk commented on tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.h:
```diff
@@ -0,0 +1,169 @@
+//    ...
+//  }
+//
+//  void profiler_collect_data_xspace(const TP_Profiler* profiler, uint8_t*
+//  buffer, size_t* size_in_bytes, TF_Status* status) {
+//    Xspace xspace = get_my_xspace(); /* Plugin generates Xspace based on
+//    collected profiler data. */ size_t buffer_size_in_bytes = *size_in_bytes;
+//    *size_in_bytes = xspace.ByteSizeLong(); /* get the size of Xspace */
+//    if (buffer == nullptr) {
+//      return; /* TensorFlow will first get the size of Xspace, then allocate
+//      the big enough buffer and pass it to the plugin for retrieving Xspace.
+//      */
+//    }
+//    bool success = xspace.SerializeToArray(buffer, buffer_size_in_bytes);
+//    }
```
Minor formatting nit:
```suggestion
//  void profiler_collect_data_xspace(const TP_Profiler* profiler, uint8_t*
//  buffer, size_t* size_in_bytes, TF_Status* status) {
//    /* Plugin generates Xspace based on collected profiler data. */
//    Xspace xspace = get_my_xspace();
//    size_t buffer_size_in_bytes = *size_in_bytes;
//    *size_in_bytes = xspace.ByteSizeLong(); /* get the size of Xspace */
//    if (buffer == nullptr) {
//      /* TensorFlow will first get the size of Xspace, then allocate the big
//         enough buffer and pass it to the plugin for retrieving Xspace. */
//      return;
//    }
//    bool success = xspace.SerializeToArray(buffer, buffer_size_in_bytes);
//  }
```

penpornk commented on tensorflow/c/experimental/pluggable_profiler/pluggable_profiler_internal.h:
```diff
@@ -0,0 +1,32 @@
+                                  TF_Status* const);
+
+Status InitPluginProfiler(TFInitProfilerFn init_fn);
+}  // namespace profiler
```
Minor formatting nit:
Please add a newline before this line.

jzhoulon replied:
done

penpornk reviewed (changes requested):
Thank you very much for the PR! 

Would you mind adding simple mock tests in a follow-up PR? You can follow examples in [stream_executor_test.cc](
https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/stream_executor/stream_executor_test.cc). In our case we'll have mock start, stop, collect_data_xspace, etc. 

For this PR, I have a few minor formatting nits.

jzhoulon committed:
Update tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.h

Co-authored-by: Penporn Koanantakool <38085909+penpornk@users.noreply.github.com>

jzhoulon committed:
Update tensorflow/c/experimental/pluggable_device_utils/BUILD

Co-authored-by: Penporn Koanantakool <38085909+penpornk@users.noreply.github.com>

jzhoulon committed:
Update tensorflow/c/experimental/pluggable_profiler/BUILD

Co-authored-by: Penporn Koanantakool <38085909+penpornk@users.noreply.github.com>

jzhoulon commented on tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.cc:
```diff
@@ -0,0 +1,42 @@
+#include "tensorflow/core/platform/regexp.h"
+#include "tensorflow/core/platform/status.h"
+#include "tensorflow/core/platform/strcat.h"
+tensorflow::Status ValidateDeviceType(tensorflow::StringPiece type) {
```
done

jzhoulon commented:
@penpornk Thanks for the comments, I have addressed all the comments, please help to review. And I will submit an follow-up PR to add mock test for pluggable profiler. Thanks.

penpornk reviewed (approved):
Thank you for changes! :)

jzhoulon committed:
address comments

jzhoulon commented:
@penpornk I saw some internal checks failed, if there is some thing need to fix, please let me know, thanks very much.

penpornk commented:
@jzhoulon Thanks for checking! I've taken care of it. Right now it's waiting for reviews from API owners. :)

jzhoulon commented on tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.h:
```diff
@@ -0,0 +1,48 @@
+namespace pluggable_device {
+
+#define VALIDATE_STRUCT_SIZE(STRUCT_NAME, STRUCT_OBJ, SIZE_VALUE_NAME)    \
+  do {                                                                    \
```
@penpornk The VALIDATA_STRUCT_SIZE seems be moved to [c_api_macros.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api_macros.h#L19-L21) after this PR merged, this will cause build failure in plugin. VALIDATE_STRUCT_SIZE is not a public C API(only used by stream_executor.cc and pluggable_profiler.cc). it contains some C++ struct(tensorflow::Status), so it will cause some issue when c_api_macros.h is included by plugin, I think this part of code needs to be moved to an TensorFlow internal files. from c_api_macros.h, Thanks very much.
build errror:
```
bazel-out/k8-opt/bin/external/local_config_tf/include/tensorflow/core/platform/status.h:287:65: note: passing argument to parameter 'v' here
inline tensorflow::string* TfCheckOpHelper(::tensorflow::Status v,
^
itex/core/graph/onednn_layout/onednn_layout.cc:263:15: error: no viable conversion from 'itex::Status' to '::tensorflow::Status'
TF_CHECK_OK(GetNodeAttr(node_def, "transpose_b", &trans_b));
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

jbaiocchi replied:
It doesn't look related. Looks like you are calling TF_CHECK_OK, which is a TF macro that expands into a function (TfCheckOpHelper) that expects ::tensorflow::Status. GetNodeAttr should return ::tensorflow::Status but somehow the compiler thinks it is itex::Status. Note that VALIDATE_STRUCT_SIZE has always depended on ::tensorflow::Status. We added ifdef __cplusplus to ensure inclusion of c_api_macros into C code is OK.

jzhoulon replied:
Thanks for the help! we find that we have duplicate name of macros in c_api_macros.h, we renamed them and it works successfully.
